Gradient-Step denoiser blog post (#207)
creatis-myriad/creatis-myriad.github.io: collections/_posts/2025-01-18-GradientStepDenoiser.md

Since applied mathematicians love optimization, this tool is itself formulated as an optimization problem:

$$\text{prox}_{\tau f} (x) = \arg \min_u f(u) + \frac{1}{2 \tau} \| x - u\|^2$$

As you can see when computing the prox of a function, the goal is to minimize the function while staying close to a given point $x$.
Long story short, there are closed form solutions for the proximal operators of a lot of usual functions.
And more precisely, most usual non-smooth regularization function that one might use in inverse problems admit closed-form proximal operators.
This is for instance the case of our previous example with the $$L_1$$ norm, but also the famous *Total Variation* regularization and others.

A typical algorithm to solve the previous minimization problem would then be the *Forward-Backward* splitting, or *Proximal Gradient Descent*.

$$x^{n+1} = \text{prox}_{\tau g} \circ \left( \text{Id} - \tau \nabla f \right) (x^{n}) $$

As you can see it's just a regular gradient descent on the smooth data-fidelity term $f$ and a proximal step on the non-smooth regularization $g$.
There are a lot of other algorithms using the proximal operator, but this is the most basic one.

---

> # Plug-and-Play

Reviewer: In this part and the following maybe state clearly which paper and contribution you are talking about.

> Let us consider an ill-posed linear inverse problem, for instance super-resolution, deblurring or inpainting. TV regularization was good looking 15 years ago, but now the reconstructions look cartoonish when compared to what end-to-end deep learning approach can achieve.

Reviewer 2: ```suggestion
Let us consider an ill-posed linear inverse problem, for instance super-resolution, deblurring or inpainting. TV regularization was good looking 15 years ago, but now the reconstructions look cartoonish when compared to what end-to-end deep learning approaches can achieve.
```

![Example with TV regularization](/collections/images/GradientStep/tv.jpg)

However reconstruction methods that use **only** neural networks also have a lot of drawbacks:
- no data-fidelity constraints, meaning that we have no theoretical guarantees on their performances

> - very sensitive to domain-shift, so applying them to datas way out of their training distribution is risky

Reviewer 2: Data is already plural :wink: 
```suggestion
- very sensitive to domain-shift, so applying them to data way out of their training distribution is risky
```

- typically require to be re-trained as soon as the degradation operator $$A$$ changes

What we really want is the best of both worlds:
- good reconstructions using neural networks
- some constraints with respect to the observations

> Plug-and-Play methods are exactly this compromise. They use the traditionnal variational formulation of inverse problems and replace the hand-crafted regularization $g$ by a Gaussian denoiser $$D_\sigma$$.

Reviewer: ```suggestion
Plug-and-Play methods are exactly this compromise. They use the traditionnal variational formulation of inverse problems and replace the hand-crafted regularization $$g$$ by a Gaussian denoiser $$D_\sigma$$.
```

> Let's take the forward-backward splitting: it's Plug-and-Play version now simply becomes:

Reviewer 2: ```suggestion
Let's take the forward-backward splitting: its Plug-and-Play version now simply becomes:
```

$$x^{n+1} = D_\sigma \circ \left( \text{Id} - \tau \nabla f \right) (x^{n})$$

These Plug-and-Play methods give very good reconstructions, with a natural robustness to domain-shift and less hallucinations than end-to-end methods.

> ![Exampleof a PnP reconstruction](/collections/images/GradientStep/pnp.jpg)

Reviewer 2: ```suggestion
![Example of a PnP reconstruction](/collections/images/GradientStep/pnp.jpg)
```

> However there are still several issues with these methods. With state-of-the-art denoisers, we have no theoretical guarantees that these schemes converge to a fix-point, let alone the minimum of our original optimization problem. 

Reviewer 2: Should it be _fixed-point_ here? I'm not sure, but I feel like it should.
```suggestion
However there are still several issues with these methods. With state-of-the-art denoisers, we have no theoretical guarantees that these schemes converge to a fixed-point, let alone the minimum of our original optimization problem. 
```

In order to have some theoretical guarantees, we need **a lot** of assumptions on $$D_\sigma$$, the most restrictive one is that $$D_\sigma$$ needs to be *contractive*, meaning that:

$$D_\sigma (x) \leq x$$

This is **very hard** to impose during training, and most of the solutions to verify this condition rely on changes in the architectures of the networks. 
This however **dramatically reduces performances**.

# Gradient-Step denoiser

> To adress these issues and recover theoretical convergence guarantees the authors formulate a special denoizer:

Reviewer 2: ```suggestion
To address these issues and recover theoretical convergence guarantees the authors formulate a special denoizer:
```

| **DRUNet**    | **40.31** | **33.97** | **31.32** | **28.08** | 69.8      |
|---------------|--------|--------|--------|--------|-----------|
| DRUNet *light*| 40.19  | 33.89  | 31.25  | 28.00  | 6.3       |
| **GS-DRUNet**     | 40.26  | 33.90  | 31.26  | 28.01  | 10.4      |

For context, DRUNet is still considered one of the best denoising model.
The authors achieve roughly similar performances while using a lighter version of the DRUnet (2 residual blocks instead of 4). They even perform slightly better than simply using the exact same architecture, without the gradient-step formulation.

# Plug-and-Play with the Gradient-Step denoiser

The interesting thing about this formulation is that now we have a regularization $$g_\sigma$$ with nice properties. More precisely it is $$L$$-smooth, which is necessary to design convergent first-order algorithms.

I'll skip the details, but the important thing is that now we can devise a **convergent Plug-and-Play** algorithm, with **any neural network** $$N_\sigma$$. The algorithm reads as follows:

$$ x^{n+1} = \text{Prox}_{\tau f} \circ \left( \text{Id} - \tau \lambda \nabla g_\sigma \right) (x^{n})$$

> This algorithm is then proven to explicitely decrease the objective function:

Reviewer 2: ```suggestion
This algorithm is then proven to explicitly decrease the objective function:
```

$$F(x) = f(x) + g_\sigma (x)$$

which was our initial objective.

> **REMARK**: you might notice that it is unusual to take a proximal step on $$f$$ and a gradient step on the regularization $$g$$. This algorithm needs to be formulated as such in order to write the convergence proof, but this is not ideal. In some cases $$f$$ does not admit a closed-form solution for its proximal operator. Subsequent work from the same authors partially fix this issue.

Reviewer 2: ```suggestion
**REMARK**: you might notice that it is unusual to take a proximal step on $$f$$ and a gradient step on the regularization $$g$$. This algorithm needs to be formulated as such in order to write the convergence proof, but this is not ideal. In some cases $$f$$ does not admit a closed-form solution for its proximal operator. Subsequent work from the same authors partially fixes this issue.
```

> # Experiments

Reviewer: I was a little lost at the end with the notation, maybe make it appears clearly in the text than GS refers to the gradient step denoiser.

The authors test their denoiser along with their new algorithm on two tasks:
- deblurring, with 10 different blur kernels (motion blur, gaussian, uniform)
- super-resolution (with gaussians)

> They perform on-par with the state-of-the-art other PnP algorithms, while offering convergence guarantees.

Reviewer: Introduce the notation PnP

Since applied mathematicians love optimization, this tool is itself formulated as an optimization problem:

$$\text{prox}_{\tau f} (x) = \arg \min_u f(u) + \frac{1}{2 \tau} \| x - u\|^2$$

As you can see when computing the prox of a function, the goal is to minimize the function while staying close to a given point $$x$$.
Long story short, there are closed form solutions for the proximal operators of a lot of usual functions.
And more precisely, most usual non-smooth regularization function that one might use in inverse problems admit closed-form proximal operators.
This is for instance the case of our previous example with the $$L_1$$ norm, but also the famous *Total Variation* regularization and others.

A typical algorithm to solve the previous minimization problem would then be the *Forward-Backward* splitting, or *Proximal Gradient Descent*.

$$x^{n+1} = \text{prox}_{\tau g} \circ \left( \text{Id} - \tau \nabla f \right) (x^{n})$$

As you can see it's just a regular gradient descent on the smooth data-fidelity term $$f$$ and a proximal step on the non-smooth regularization $$g$$.
There are a lot of other algorithms using the proximal operator, but this is the most basic one.

> Interested readers can refer to the textbook [1] for a complete overview of these methods.

Reviewer 2: ```suggestion
Interested readers can refer to the textbook [^1] for a complete overview of these methods.
```

# Plug-and-Play

---

Let us consider an ill-posed linear inverse problem, for instance super-resolution, deblurring or inpainting. TV regularization was good looking 15 years ago, but now the reconstructions look cartoonish when compared to what end-to-end deep learning approaches can achieve.

![Example with TV regularization](/collections/images/GradientStep/tv.jpg)

However reconstruction methods that use **only** neural networks also have a lot of drawbacks:
- no data-fidelity constraints, meaning that we have no theoretical guarantees on their performances
- very sensitive to domain-shift, so applying them to data way out of their training distribution is risky
- typically require to be re-trained as soon as the degradation operator $$A$$ changes

What we really want is the best of both worlds:
- good reconstructions using neural networks
- some constraints with respect to the observations

> Plug-and-Play methods [2] are exactly this compromise. They use the traditionnal variational formulation of inverse problems and replace the hand-crafted regularization $$g$$ by a Gaussian denoiser $$D_\sigma$$.

Reviewer 2: ```suggestion
Plug-and-Play methods [^2] are exactly this compromise. They use the traditionnal variational formulation of inverse problems and replace the hand-crafted regularization $$g$$ by a Gaussian denoiser $$D_\sigma$$.
```

Let's take the forward-backward splitting: its Plug-and-Play version now simply becomes:

$$x^{n+1} = D_\sigma \circ \left( \text{Id} - \tau \nabla f \right) (x^{n})$$

> These Plug-and-Play methods give very good reconstructions [3], with a natural robustness to domain-shift and less hallucinations than end-to-end methods.

Reviewer 2: ```suggestion
These Plug-and-Play methods give very good reconstructions [^3], with a natural robustness to domain-shift and less hallucinations than end-to-end methods.
```

![Example of a PnP reconstruction](/collections/images/GradientStep/pnp.jpg)

However there are still several issues with these methods. With state-of-the-art denoisers, we have no theoretical guarantees that these schemes converge to a fixed-point, let alone the minimum of our original optimization problem.

> In order to have some theoretical guarantees, we need **a lot** of assumptions on $$D_\sigma$$ [4], the most restrictive one is that $$D_\sigma$$ needs to be *contractive*, meaning that:

Reviewer 2: ```suggestion
In order to have some theoretical guarantees, we need **a lot** of assumptions on $$D_\sigma$$ [^4], the most restrictive one is that $$D_\sigma$$ needs to be *contractive*, meaning that:
```

$$D_\sigma (x) \leq x$$

> This is **very hard** to impose during training, and most of the solutions to verify this condition rely on changes in the architectures of the networks [5].

Reviewer 2: ```suggestion
This is **very hard** to impose during training, and most of the solutions to verify this condition rely on changes in the architectures of the networks [^5].
```

The authors achieve roughly similar performances while using a lighter version of the DRUnet (2 residual blocks instead of 4). They even perform slightly better than simply using the exact same architecture, without the gradient-step formulation.

# Plug-and-Play with the Gradient-Step denoiser

The interesting thing about this formulation is that now we have a regularization $$g_\sigma$$ with nice properties. More precisely it is $$L$$-smooth, which is necessary to design convergent first-order algorithms.

I'll skip the details, but the important thing is that now we can devise a **convergent Plug-and-Play** algorithm, with **any neural network** $$N_\sigma$$. The algorithm reads as follows:

$$ x^{n+1} = \text{Prox}_{\tau f} \circ \left( \text{Id} - \tau \lambda \nabla g_\sigma \right) (x^{n})$$

which is pretty similar to the standard Forward-Backward algorithm. 
Replacing the gradient step on $$g_\sigma$$ by its formulation as a denoiser gives:

$$ x^{n+1} = \text{Prox}_{\tau f} \circ D_\sigma (x^{n})$$

This algorithm is then proven to explicitly decrease the objective function:

$$F(x) = f(x) + g_\sigma (x)$$

which was our initial objective.

> **REMARK**: you might notice that it is unusual to take a proximal step on $$f$$ and a gradient step on the regularization $$g$$. This algorithm needs to be formulated as such in order to write the convergence proof, but this is not ideal. In some cases $$f$$ does not admit a closed-form solution for its proximal operator. Subsequent work from the same authors partially fixes this issue [6-7].

Reviewer 2: ```suggestion
**REMARK**: you might notice that it is unusual to take a proximal step on $$f$$ and a gradient step on the regularization $$g$$. This algorithm needs to be formulated as such in order to write the convergence proof, but this is not ideal. In some cases $$f$$ does not admit a closed-form solution for its proximal operator. Subsequent work from the same authors partially fixes this issue [^6][^7].
```

![Super-resolution experiments](/collections/images/GradientStep/superresolution.jpg)

The extensive experiments of the authors show that they perform on-par with SOTA methods while using a lighter denoiser.

![Quantitative results on the deblurring task](/collections/images/GradientStep/results.jpg)

# Conclusion

Why does this matter ?

- state-of-the-art performances
- a single network for a wide variety of image restoration problems
- now we can confidently use deep learning models in image restoration
- no need to select the "best" iteration, or use early-stopping: the algorithm is convergent !
- a new framework to easily design other convergent Plug-and-Play algorithms for inverse problems
- opens up a lot of new research directions

# References

[1] First-Order Methods in Optimization, Amir Beck, 2017

[2] Plug-and-Play priors for model based reconstruction, Venkatakrishnan et al., 2013

[3] Plug-and-Play Image Restoration With Deep Denoiser Prior, Zhang et al., 2022

[4] Plug-and-Play Methods Provably Converge with Properly Trained Denoisers, Ryu et al., 2019

> [5] Plug-and-Play Methods Provably Converge with Properly Trained Denoisers, Terris et al., 2020

Reviewer 2: The titles of both references are the same, which I suspect was a copy-paste issue :wink:

> [1] First-Order Methods in Optimization, Amir Beck, 2017
> 
> [2] Plug-and-Play priors for model based reconstruction, Venkatakrishnan et al., 2013
> 
> [3] Plug-and-Play Image Restoration With Deep Denoiser Prior, Zhang et al., 2022
> 
> [4] Plug-and-Play Methods Provably Converge with Properly Trained Denoisers, Ryu et al., 2019
> 
> [5] Plug-and-Play Methods Provably Converge with Properly Trained Denoisers, Terris et al., 2020
> 
> [6] Proximal denoiser for convergent plug-and-play optimization with nonconvex regularization, Hurault et al., 2022
> 
> [7] Convergent Bregman Plug-and-Play Image Restoration for Poisson Inverse Problems, Hurault et al., 2023

Reviewer 2: ```suggestion
[^1]: First-Order Methods in Optimization, Amir Beck, 2017
[^2]: Plug-and-Play priors for model based reconstruction, Venkatakrishnan et al., 2013
[^3]: Plug-and-Play Image Restoration With Deep Denoiser Prior, Zhang et al., 2022
[^4]: Plug-and-Play Methods Provably Converge with Properly Trained Denoisers, Ryu et al., 2019
[^5]: Plug-and-Play Methods Provably Converge with Properly Trained Denoisers, Terris et al., 2020
[^6]: Proximal denoiser for convergent plug-and-play optimization with nonconvex regularization, Hurault et al., 2022
[^7]: Convergent Bregman Plug-and-Play Image Restoration for Poisson Inverse Problems, Hurault et al., 2023
```
